### DISCUSSION TITLE
Cannot Alter values of Vi.Mock for different test cases

### DISCUSSION THREAD

**Discussion by shreyan1993** [vitest-dev/vitest#5676] [Q&A]
### Describe the bug

Hi Guys, 

i am trying to use vi mock for the below code, but i am unable to alter the value of isProduction . It is always defaulting to the first case and i cannot mockresolvevalue as its not really a function.

`import { ThemeProvider } from '@mui/material/styles';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { IEnvironment } from 'src/interfaces';
import { lang } from 'src/utils';
import { theme } from 'src/utils/theme';
import { vi } from 'vitest';

import { About } from './About';

describe('About Component', () => {
  beforeEach(() => {
    vi.mock('src/hooks', () => ({
      useGetProjectConfig: () => ({
        data: {
          gcf: {
            apps: {
              player: {
                version: '1.0.0',
              },
              cod: {
                version: '2.0.0',
              },
              components: {
                version: '3.0.0',
              },
            },
          },
        },
      }),
    }));

    window.environment = {
      urls: {
        ccBeApi: 'https://cc-be-2-1-0-rc20240207-48f3e1c4-au.coates.io',
      },
    } as IEnvironment;
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.resetModules();
  });

  it('renders about component with correct versions when isProduction is true', () => {
    vi.mock('src/utils/environment', () => ({
      isProduction: true,
      SB_API_URL: '',
      CCBE_API_URL: '',
    }));

    render(
      <ThemeProvider theme={theme}>
        <About />
      </ThemeProvider>,
    );

    expect(screen.getByText(lang.ccUI)).toBeInTheDocument();
    expect(screen.getByText('1.0.0')).toBeInTheDocument();
    expect(screen.getByText(lang.ccBE)).toBeInTheDocument();
    expect(screen.getByText('2.1.0')).toBeInTheDocument();
    expect(screen.getByText(lang.gcfPlayer)).toBeInTheDocument();
    expect(screen.getByText('1.0.0')).toBeInTheDocument();
    expect(screen.getByText(lang.gcfCOD)).toBeInTheDocument();
    expect(screen.getByText('2.0.0')).toBeInTheDocument();
    expect(screen.getByText(lang.gcfComponents)).toBeInTheDocument();
    expect(screen.getByText('3.0.0')).toBeInTheDocument();
  });

  it('renders about component with correct versions when isProduction is false', () => {
    vi.resetModules();
    vi.mock('src/utils/environment', () => ({
      isProduction: false,
      SB_API_URL: '',
      CCBE_API_URL: '',
    }));

    render(
      <ThemeProvider theme={theme}>
        <About />
      </ThemeProvider>,
    );

    expect(screen.getByText(lang.ccBE)).toBeInTheDocument();
    expect(screen.getByText('2.1.0 RC 2024/02/07')).toBeInTheDocument();
  });
});
`

### Reproduction

`import { ThemeProvider } from '@mui/material/styles';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { IEnvironment } from 'src/interfaces';
import { lang } from 'src/utils';
import { theme } from 'src/utils/theme';
import { vi } from 'vitest';

import { About } from './About';

describe('About Component', () => {
  beforeEach(() => {
    vi.mock('src/hooks', () => ({
      useGetProjectConfig: () => ({
        data: {
          gcf: {
            apps: {
              player: {
                version: '1.0.0',
              },
              cod: {
                version: '2.0.0',
              },
              components: {
                version: '3.0.0',
              },
            },
          },
        },
      }),
    }));

    window.environment = {
      urls: {
        ccBeApi: 'https://cc-be-2-1-0-rc20240207-48f3e1c4-au.coates.io',
      },
    } as IEnvironment;
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.resetModules();
  });

  it('renders about component with correct versions when isProduction is true', () => {
    vi.mock('src/utils/environment', () => ({
      isProduction: true,
      SB_API_URL: '',
      CCBE_API_URL: '',
    }));

    render(
      <ThemeProvider theme={theme}>
        <About />
      </ThemeProvider>,
    );

    expect(screen.getByText(lang.ccUI)).toBeInTheDocument();
    expect(screen.getByText('1.0.0')).toBeInTheDocument();
    expect(screen.getByText(lang.ccBE)).toBeInTheDocument();
    expect(screen.getByText('2.1.0')).toBeInTheDocument();
    expect(screen.getByText(lang.gcfPlayer)).toBeInTheDocument();
    expect(screen.getByText('1.0.0')).toBeInTheDocument();
    expect(screen.getByText(lang.gcfCOD)).toBeInTheDocument();
    expect(screen.getByText('2.0.0')).toBeInTheDocument();
    expect(screen.getByText(lang.gcfComponents)).toBeInTheDocument();
    expect(screen.getByText('3.0.0')).toBeInTheDocument();
  });

  it('renders about component with correct versions when isProduction is false', () => {
    vi.resetModules();
    vi.mock('src/utils/environment', () => ({
      isProduction: false,
      SB_API_URL: '',
      CCBE_API_URL: '',
    }));

    render(
      <ThemeProvider theme={theme}>
        <About />
      </ThemeProvider>,
    );

    expect(screen.getByText(lang.ccBE)).toBeInTheDocument();
    expect(screen.getByText('2.1.0 RC 2024/02/07')).toBeInTheDocument();
  });
});
`

### System Info

```shell
Mac OS
```

### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ashokmn1** [comment]
`vi.mock` calls are hoisted to the top of the file and only run once, so calling it inside `it()` or `beforeEach()` won't change the mock between tests. That's why it always uses the first value.

You can fix this with `vi.hoisted` to create a mutable object that the mock reads from:

```tsx
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import { About } from './About';

const env = vi.hoisted(() => ({ isProduction: true }));

vi.mock('src/utils/environment', () => ({
  get isProduction() {
    return env.isProduction;
  },
  SB_API_URL: '',
  CCBE_API_URL: '',
}));

describe('About Component', () => {
  it('renders correct versions when isProduction is true', () => {
    env.isProduction = true;
    render(<About />);
    // assertions
  });

  it('renders correct versions when isProduction is false', () => {
    env.isProduction = false;
    render(<About />);
    // assertions
  });
});
```

The getter is the important part: it makes your component read the current value each time it's accessed, instead of the value captured when the module was first imported.

Also move your `src/hooks` mock out of `beforeEach` to the top level, since it's hoisted anyway.

**Alternative:** if you need a completely fresh module per test, use `vi.doMock` (not hoisted) with a dynamic import:

```tsx
it('renders when isProduction is false', async () => {
  vi.resetModules();
  vi.doMock('src/utils/environment', () => ({
    isProduction: false,
    SB_API_URL: '',
    CCBE_API_URL: '',
  }));
  const { About } = await import('./About');
  render(<About />);
  // assertions
});
```

Docs: [vi.hoisted](https://vitest.dev/api/vi.html#vi-hoisted) · [vi.doMock](https://vitest.dev/api/vi.html#vi-domock)
